### PR TITLE
NRS combined BGC product + larval fish

### PR DESCRIPTION
**Pull request by leonardolaiolo**

### CODE REVIEW

**Comment by leonardolaiolo** [comment]
Update to @ mhidas :
Marty, this branch is okay to pull and test. Unfortunately I did not manage to progress significantly here, I believe modification are still needed in the ```yaml``` files to create tables or in the .csv files. Here a summary of my work:
- ```bgc_pigments.dataresource.yaml``` changes I made here allow to produce the ```bgc_pigments``` table
- ```bgc_tss.dataresource.yaml``` changes I made here allow to produce the ```bgc_tss``` table
- ```bgc_trip.dataresource.yaml``` this did not need changes as the ```bgc_trip``` table was already successfully produced earlier
From the Combined BGC product specification document we need another 2 tables:
- ```bgc_chemistry``` 
- ```bgc_picoplankton``` 
The harvester fails on creating both tables from

File: bgc_data/bgc_chemistry.dataresource.yaml [hunk 1]
```diff
@@ -10,25 +10,14 @@ layout:
...
   - name: TRIP_CODE
     title: "Identifier for trip"
     type: string
     required: true
   - name: SAMPLEDEPTH_M
     title: "Depth (m) that sample was obtained"
-    type: integer
+    type: number
     required: true
-  - name: SAMPLEDATELOCAL
```
[the remark this answers is not in the document]
Comment (leonardolaiolo) [reply bgc_data/bgc_chemistry.dataresource.yaml:25]: fixed in [this commit](https://github.com/aodn/public-schema/pull/15/commits/002e078b63113ce1a44197928d6f636f9a346ed0)

File: bgc_data/bgc_chemistry.dataresource.yaml [hunk 2]
```diff
@@ -128,8 +105,8 @@ schema:
   primaryKey:
   - TRIP_CODE
   - SAMPLEDEPTH_M
-  - REPLICATE
```
[the remark this answers is not in the document]
Comment (leonardolaiolo) [reply bgc_data/bgc_chemistry.dataresource.yaml:131]: fixed in [this commit](https://github.com/aodn/public-schema/pull/15/commits/002e078b63113ce1a44197928d6f636f9a346ed0)

File: bgc_data/bgc_picoplankton.dataresource.yaml [hunk 3]
```diff
@@ -20,63 +20,29 @@ schema:
     required: true
   - name: SAMPLEDEPTH_M
     title: "Depth (m) that sample was obtained, contains Flag Value"
-    type: number
+    type: string
     required: true    
-  - name: SAMPLEDATELOCAL
```
[the remark this answers is not in the document]
Comment (leonardolaiolo) [reply bgc_data/bgc_picoplankton.dataresource.yaml:25]: fixed in [this commit](https://github.com/aodn/public-schema/pull/15/commits/002e078b63113ce1a44197928d6f636f9a346ed0)

File: bgc_data/bgc_picoplankton.dataresource.yaml [hunk 4]
```diff
@@ -119,24 +82,22 @@ schema:
...
-    title: ""
     type: integer                               
     required: false  
   - name: BATCH_COMMENTS
     title: "Optional comments on batch process"
     type: string                               
     required: false
   primaryKey:
-  - TAXON_NAME
+  - SAMPLE_CODE
```
[the remark this answers is not in the document]
Comment (leonardolaiolo) [reply bgc_data/bgc_picoplankton.dataresource.yaml]: fixed in [this commit](https://github.com/aodn/public-schema/pull/15/commits/002e078b63113ce1a44197928d6f636f9a346ed0)

File: bgc_data/bgc_pigments.dataresource.yaml [hunk 5]
```diff
@@ -10,201 +10,145 @@ layout:
...
-    title: "Source project name"
-    type: string
-    required: false
-    unique: false
   - name: TRIP_CODE
     title: "Identifier for trip"
     type: string
     required: true
     unique: true
-  - name: SAMPLEDATELOCAL
```
[the remark this answers is not in the document]
Comment (leonardolaiolo) [reply bgc_data/bgc_pigments.dataresource.yaml:23]: fixed in [this commit](https://github.com/aodn/public-schema/pull/15/commits/002e078b63113ce1a44197928d6f636f9a346ed0)

File: bgc_data/bgc_tss.dataresource.yaml [hunk 6]
```diff
@@ -14,13 +14,6 @@ schema:
     title: "Identifier for trip"
     type: string
     required: true
-  - name: SAMPLEDATELOCAL
```
[the remark this answers is not in the document]
Comment (leonardolaiolo) [reply bgc_data/bgc_tss.dataresource.yaml:17]: fixed in [this commit](https://github.com/aodn/public-schema/pull/15/commits/002e078b63113ce1a44197928d6f636f9a346ed0)

File: bgc_data/bgc_tss.dataresource.yaml [hunk 7]
```diff
@@ -30,27 +23,21 @@ schema:
...
-    title: ""
     type: number                               
     required: false  
   - name: ORGANICFRACTION_MGL
-    title: ""
     type: number                               
     required: false 
   - name: TSS_FLAG
-    title: ""
     type: string
```
[the remark this answers is not in the document]
Comment (leonardolaiolo) [reply bgc_data/bgc_tss.dataresource.yaml]: fixed in [this commit](https://github.com/aodn/public-schema/pull/15/commits/002e078b63113ce1a44197928d6f636f9a346ed0)

**Comment by leonardolaiolo** [comment]
@mhidas @clairedavies 
I have implemented the changes, thus I believe this branch can be merged to master?

File: bgc_data/bgc_chemistry.dataresource.yaml [hunk 8]
```diff
@@ -21,10 +21,10 @@ schema:
   - name: SAMPLEDATELOCAL
     title: "Date & time of sample in local time (formatted string)"
     type: datetime
-    # format string *must* be specified using standard Python/C strptime specification         
-    # https://docs.python.org/2/library/datetime.html#strftime-strptime-behavior               
+    # format string *must* be specified using standard Python/C strptime specification
+    # https://docs.python.org/2/library/datetime.html#strftime-strptime-behavior
     format: "%Y-%m-%d %H:%M:%S"
-    required: false
+    required: true
```
Comment (mhidas) [inline bgc_data/bgc_chemistry.dataresource.yaml:27]: @clairedavies If I remember correctly you said in the catch-up meeting that these sample date fields should always be populated.
Same in all these tables?

File: bgc_data/bgc_lfish_samples.dataresource.yaml [hunk 9]
```diff
@@ -15,10 +15,6 @@ schema:
     type: integer
     required: true
     unique: true
-  - name: PROJECTNAME
-    title: "Source project name"
-    type: string
-    required: true
```
Comment (mhidas) [inline bgc_data/bgc_lfish_samples.dataresource.yaml:21]: As with the other tables, PROJECTNAME, as well as LATITUDE, LONGITUDE, and STATIONNAME (below) are already in the `bgc_trip` table, so don't need to be included here.
@clairedavies This is assuming the Larval Fish products are all from (a subset of?) the same trips?

File: bgc_data/bgc_picoplankton.dataresource.yaml [hunk 10]
```diff
@@ -102,11 +102,10 @@ schema:
     type: string                               
     required: false
   primaryKey:
-  - SAMPLEDEPTH_M
   - TRIP_CODE
-  - REPLICATE   
+  - SAMPLEDEPTH_M
+  - REPLICATE
```
Comment (mhidas) [inline bgc_data/bgc_picoplankton.dataresource.yaml:107]: This order just makes more sense to me.

File: bgc_data/bgc_pigments.dataresource.yaml [hunk 11]
```diff
@@ -18,14 +18,14 @@ schema:
...
+    # https://docs.python.org/2/library/datetime.html#strftime-strptime-behavior
     format: "%Y-%m-%d %H:%M:%S"
-    required: false
+    required: true
   - name: SAMPLEDEPTH_M
-    title: "Depth (m) that sample was obtained"
+    title: "Depth (m) that sample was obtained, or 'WC' if combined water-column sample"
     type: string
-    required: false
+    required: true
```
Comment (mhidas) [inline bgc_data/bgc_pigments.dataresource.yaml:28]: If SAMPLEDEPTH_M is part of the primary key, it must be required

**Review by mhidas — COMMENTED** [review]
@clairedavies @sedgar A few more suggestions to review. 
Some of these (if accepted) will require Geoserver changes too.

File: bgc_data/bgc_chemistry.dataresource.yaml (hunk 8 above)
Comment (clairedavies) [reply bgc_data/bgc_chemistry.dataresource.yaml:27]: Yes, every sample has a date field

File: bgc_data/bgc_lfish_samples.dataresource.yaml (hunk 9 above)
Comment (clairedavies) [reply bgc_data/bgc_lfish_samples.dataresource.yaml:21]: The larval fish data includes non NRS data and so this information is required as they are NOT all from a subset of the same trips. There is some voyage information here that doesn't apply to the NRS. Where projectname = IMOS_NRS_xxx then the data is in the bgc_trip table, where projectname != IMOS_NRS_xxx then this information is not in the bgc_trip table.

File: bgc_data/bgc_picoplankton.dataresource.yaml (hunk 10 above)
Comment (clairedavies) [reply bgc_data/bgc_picoplankton.dataresource.yaml:107]: OK

File: bgc_data/bgc_pigments.dataresource.yaml (hunk 11 above)
Comment (clairedavies) [reply bgc_data/bgc_pigments.dataresource.yaml:28]: agree

File: bgc_data/bgc_lfish_samples.dataresource.yaml (hunk 9 above)
Comment (leonardolaiolo) [reply bgc_data/bgc_lfish_samples.dataresource.yaml:21]: PROJECTNAME, LAT, LON and STATIONNAME add on [this commit](https://github.com/aodn/public-schema/pull/15/commits/4435f213a4cd7f41af3840167afa0b8cf7185cf8)

Comment (mhidas) [reply bgc_data/bgc_lfish_samples.dataresource.yaml:21]: Thanks for clarifying @clairedavies. I'm happy with that, if you think that's the easiest way to manage these data.
Alternatively, couldn't we just add all the Larval Fish trips into the `bgc_trip` table? There are already plenty of non-NRS trips in there!
Or is that more complicated at your end?

**Comment by mhidas** [comment]
@clairedavies I think the schemas for the BGC product are pretty good now.

@sedgar The Geoserver layers for `bgc_chemistry` and `bgc_pigments` are currently incosistent with these schemas and will need to be updated (see [test results](https://github.com/aodn/public-schema/runs/4147672624?check_suite_focus=true#step:5:15) - I think it's just the PROJECTNAME column that needs to be removed).

In addition, it looks like `bgc_pigments` and `bgc_tss` do have rows with a null in the SAMPLEDATELOCAL column. (Annoyingly the validation checks are not picking this up - I'm not sure why.)

**Comment by clairedavies** [comment]
Missing dates are sorted, will look at the schemas for BGC products, but I think from what I saw previously that they will be OK.

Steve and I will sort these other issues when we are both back on deck in early Dec. I have a list of things we need to do. Hopefully there are no show stoppers in the short term.

Claire

File: bgc_data/bgc_lfish_samples.dataresource.yaml (hunk 9 above)
Comment (clairedavies) [reply bgc_data/bgc_lfish_samples.dataresource.yaml:21]: I prefer it that way as a) they are voyage data not station data, b) they are only relevant to fish, nothing else, c) the fish can then be stand alone as a product for all their data which is what Iain wants, d) they are finishing soon and therefore once that is done we can move on.

**Comment by clairedavies** [comment]
> In addition, it looks like `bgc_pigments` and `bgc_tss` do have rows with a null in the SAMPLEDATELOCAL column. (Annoyingly the validation checks are not picking this up - I'm not sure why.)

Sorted now, all have dates, this was my mistake

**Review by clairedavies — APPROVED** [review]

**Comment by mhidas** [comment]
Thanks @clairedavies , tss and pigments look better now!

> Steve and I will sort these other issues when we are both back on deck in early Dec. I have a list of things we need to do. Hopefully there are no show stoppers in the short term.

Yep, that'll be fine. We'll just apply those changes in local copies and work off them until then.

**Pull request merged**